File: singleton/Singleton.java
public class SupaDatabase { 

    private SupaDatabase() {
        // Private constructor to prevent instantiation
    }

    /* Java loads static classes lazily (only when needed)
    * new SupaDatabaseHolder() is not called until getInstance() is called
    * also it is called only once during loading of the class
    * in antother place getInstance() is called SupaBaseHolder is already loaded
    */
    private static class SupaDatabaseHolder {
        private static final SupaDatabase INSTANCE = new SupaDatabase();
    }

    public static SupaDatabase getInstance() {
        return SupaDatabaseHolder.INSTANCE;
    }

    // Other methods for database operations can be added here
}

File: singleton/Singleton.kt
class SupaDatabase private constructor() { // no one can create an instance directly

    init { // runs only once
        // Initialize the database connection or any other setup
    }

    companion object {
        @Volatile // ensures visibility of changes to the instance across threads
        private var instance: SupaDatabase? = null

        fun getInstance(): SupaDatabase {
            return instance ?: synchronized(this) { // makes it thread-safe during creation
                instance ?: SupaDatabase().also { instance = it }
            }
        }
    }
}